### ISSUE TITLE
[BUG] GKE hosted cluster upgrade options involving `upgrade node version` produce inconsistent results

### ISSUE THREAD

**Issue by slickwarren** [rancher/rancher#40983]
**Rancher Server Setup**
- Rancher version: v2.7.2-rc7
- Installation option (Docker install/Helm Chart): helm
   - If Helm Chart, Kubernetes Cluster and version (RKE1, RKE2, k3s, EKS, etc): rke2
- Proxy/Cert Details: valid, signed certs

**Information about the Cluster**
- Kubernetes version: v1.25.6-gke.1000 -> v1.25.7-gke.1000
- Cluster Type (Local/Downstream): downstream
   - If downstream, what type of cluster? (Custom/Imported or specify provider for Hosted/Infrastructure Provider): GKE hosted cluster

**User Information**
- What is the role of the user logged in? (Admin/Cluster Owner/Cluster Member/Project Owner/Project Member/Custom)
  - If custom, define the set of permissions: standard user, cluster owner

**Describe the bug**

when upgrading hosted clusters, there is an option to upgrade the node versions along with the cluster upgrade. However, there appear to be the following issues when using this option in GKE:
* selecting the option will upgrade the cluster, but not the node versions
* not selecting this option puts the cluster in an upgrading state, however after completion the previous version persists.

**To Reproduce**

1 deploy a GKE hosted cluster through rancher UI on v1.25.6-gke.1000
2 wait for the cluster to get to an active state
3a upgrade the cluster **without** checking the `upgrade node version` option from the `additional options` section
  * observe that the k8s version of the cluster does _not_  upgrade
3b proceed with steps 1 and 2 on a fresh cluster, then upgrade, this time selecting the `upgrade node version` option
  * observe that the cluster version _may be_ upgraded, but the nodes are not upgraded
  * on further investigation, this appears to not upgrade cluster version for patch version upgrades, but does upgrade minor versions for the cluster. Neither result in upgrading the node versions
3c proceed with steps 1, 2, and 3a on a fresh cluster, then step 3b.  
  * observe that the cluster version will not upgrade, but the node versions are upgraded

**Result**
see observation notes above

**Expected Result**

not selecting the `upgrade node version` option should allow the cluster to upgrade
selecting the `upgrade node version` should upgrade both the cluster and the node versions. 

**Screenshots**

<img width="572" alt="Screen Shot 2023-03-24 at 10 33 06 PM" src="https://user-images.githubusercontent.com/16691014/227699459-4ef2bcba-9e9e-44a9-9c17-9a5778db2fa2.png">

**Additional Information**
* Some of the above cases may be intermittent. I noticed toward the end of my investigation that sometimes, the upgrade would silently fail altogether. i.e., selecting the `upgrade node versions` would trigger an upgrade and show the updated version in rancher UI, however after waiting for it to complete, would switch back to the previous version and go back to an active state. 
**Additional Info**
* this appears to be a regression and does not happen on v2.7.1 when upgrading from 1.23.16 -> 1.24.9

**Comment by richardcase** [comment]
Tested the following:

- Using Rancher **v2.7.2-rc7** created a GKE cluster with version `v1.25.6-gke.1000` for control plane and node group (ng had 1 node initially)
- Wait for the cluster to be **Active**
- Using Rancher i edited the config for the cluster and changed the control plane version to be `v1.25.7-gke.1000` and selected the `Upgrade Node version....` option
- This resulted in this request body:
```json
{"name":"richtest1","gkeConfig":{"ipAllocationPolicy":{},"kubernetesVersion":"1.25.7-gke.1000","locations":[],"nodePools":[{"autoscaling":{"enabled":false,"maxNodeCount":0,"minNodeCount":0,"type":"/v3/schemas/gkeNodePoolAutoscaling"},"config":{"diskSizeGb":100,"localSsdCount":0,"preemptible":false,"diskType":"pd-standard","imageType":"COS_CONTAINERD","machineType":"n1-standard-2","type":"/v3/schemas/gkeNodeConfig","oauthScopes":["https://www.googleapis.com/auth/devstorage.read_only","https://www.googleapis.com/auth/logging.write","https://www.googleapis.com/auth/monitoring","https://www.googleapis.com/auth/servicecontrol","https://www.googleapis.com/auth/service.management.readonly","https://www.googleapis.com/auth/trace.append"],"labels":{}},"initialNodeCount":1,"management":{"autoRepair":true,"autoUpgrade":true,"type":"/v3/schemas/gkeNodePoolManagement"},"maxPodsConstraint":110,"name":"ng1","type":"/v3/schemas/gkeNodePoolConfig","version":"1.25.7-gke.1000"}]},"annotations":{"authz.management.cattle.io/creator-role-bindings":"{\"created\":[\"cluster-owner\"],\"required\":[\"cluster-owner\"]}","clusters.management.cattle.io/ke-last-refresh":"1679990527","lifecycle.cattle.io/create.cluster-agent-controller-cleanup":"true","lifecycle.cattle.io/create.cluster-provisioner-controller":"true","lifecycle.cattle.io/create.cluster-scoped-gc":"true","lifecycle.cattle.io/create.mgmt-cluster-rbac-remove":"true","management.cattle.io/current-cluster-controllers-version":"v1.25.6-gke.1000"}}
```
Result: control plane and node group were updated as expected.

**Comment by richardcase** [comment]
Run another test with the following:

- Upgrade `v1.25.6-gke.1000` to `v1.25.7-gke.1000` including `Upgrade Node version....` = both cp and nodes upgraded as expected
- Upgrade `1.24.11-gke.1000` to `1.25.6-gke.1000`  including `Upgrade Node version....` = both cp and nodes upgraded as expected

**Comment by richardcase** [comment]
Another test:

- Upgrade `1.25.6-gke.1000`  to `1.25.7-gke.1000 (without node)` **without** `Upgrade Node version....` = just cp is updated as expected

**Comment by richardcase** [comment]
And confirming Rancher `v2.7.2-rc7`:

![image](https://user-images.githubusercontent.com/198425/228221336-85b287b0-a4d6-4bc0-9666-a9bd659bbead.png)

**Comment by richardcase** [comment]
Tried the Rancher upgrade scenario and did the following:

- Install Rancher **2.7.1** first
- `helm install rancher rancher-latest/rancher --namespace cattle-system --set bootstrapPassword=admin --set replicas=1 --set hostname=tun.fruitcase.dev --version 2.7.1`
- Create GKE cluster version **1.24.11-gke.1000** (highest supported version via the UI)
- Wait for cluster active
- Check what the gke-operator version (rancher/gke-operator:v1.1.5 in my case)
- Upgrade Rancher to **2.7.2-rc7**
- `helm upgrade rancher rancher-latest/rancher --namespace cattle-system --set bootstrapPassword=admin --set replicas=1 --set hostname=tun.fruitcase.dev --version 2.7.2-rc7`
- Wait for the upgrades to finish (including all the helm operations)
- Reload the Rancher UI (ensure version is **2.7.2-rc7** in bottom left corner)
- Check gke-operator version again (still rancher/gke-operator:v1.1.5)
- Edit config of GKE cluster
- Upgrade k8s version in **Cluster Options** to 1.25.6-gke.1000 and in **Node Pools** tick `Upgrade Node Version from 1.24.11-gke.1000 to 1.25.6-gke.1000`
- Request body:
	  ```json
	  {"name":"richtest5","gkeConfig":{"ipAllocationPolicy":{},"kubernetesVersion":"1.25.6-gke.1000","locations":[],"nodePools":[{"autoscaling":{"enabled":false,"maxNodeCount":0,"minNodeCount":0,"type":"/v3/schemas/gkeNodePoolAutoscaling"},"config":{"diskSizeGb":100,"localSsdCount":0,"preemptible":false,"diskType":"pd-standard","imageType":"COS_CONTAINERD","machineType":"n1-standard-2","type":"/v3/schemas/gkeNodeConfig","oauthScopes":["https://www.googleapis.com/auth/devstorage.read_only","https://www.googleapis.com/auth/logging.write","https://www.googleapis.com/auth/monitoring","https://www.googleapis.com/auth/servicecontrol","https://www.googleapis.com/auth/service.management.readonly","https://www.googleapis.com/auth/trace.append"],"labels":{}},"initialNodeCount":1,"management":{"autoRepair":true,"autoUpgrade":true,"type":"/v3/schemas/gkeNodePoolManagement"},"maxPodsConstraint":110,"name":"ng1","type":"/v3/schemas/gkeNodePoolConfig","version":"1.25.6-gke.1000"}]}}
	  ```
- Wait for cluster to upgrade (check gke console and Rancher)
- **Result = cluster (a.k.a control plane) upgraded as expected, node pool upgraded as expected**
- Then upgrade cluster to **1.25.7-gke.1000** using the same method
- Wait for cluster to upgrade (check gke console and Rancher)
- **Result = cluster (a.k.a control plane) upgraded as expected, node pool upgraded as expected**

**Comment by richardcase** [comment]
@slickwarren - i haven't been able to repro this issue even with the Rancher upgrade 2.7.1 -> 2.7.2-rc7.

**Comment by slickwarren** [comment]
hmm, I was able to get a similar result on my first attempt:

* on rancher 2.7.1 (local cluster: 3 node, rke1 v1.4.0) deploy a 3 cluster GKE 1.23.16-gke.1400 (I was in us-west2b)
* wait for the cluster to get to an active state
* upgrade rancher to v2.7.2-rc7
* upgrade GKE to latest 1.24 available (currently 1.24.11-gke.1000) 
  * also tick the `upgrade nodes` box near the end of the screen before upgrading the cluster
* wait for the cluster to upgrade

- user gets an error during the upgrade (the first time around, there was no error seen in the UI


```
googleapi: Error 400: Master cannot be upgraded to "1.23.16-gke.1400": specified version is not newer than the current version., badRequest

```
<img width="968" alt="Screen Shot 2023-03-30 at 4 29 15 PM" src="https://user-images.githubusercontent.com/16691014/228986946-82c22128-ea69-4c70-8aa0-e08e6424e0a4.png">

@richardcase I have a setup you can use if you ping me directly.

**Comment by slickwarren** [comment]
I was unable to reproduce a 3rd time, this time upgrading to v2.7.2-rc8. Closing and signing off for now.
